File: src/server/gamestore.ts
import type { Game } from '../shared/schema.js'
import { saveJson, loadJson } from './storage.js'

const store = new Map<string, Game>()

export function getGame(id: string): Game | undefined {
  return store.get(id)
}

export function setGame(game: Game): void {
  store.set(game.id, game)
}

export async function getOrLoadGame(id: string): Promise<Game | null> {
  if (store.has(id)) return store.get(id)!
  const loaded = await loadJson<Game>(`games/${id}/state.json`)
  if (loaded) store.set(id, loaded)
  return loaded
}

export async function snapshotGame(game: Game): Promise<void> {
  await saveJson(`games/${game.id}/state.json`, game)
}

export function allGames(): Game[] {
  return Array.from(store.values())
}

File: server/gamestore.py
from server.schema import Game
from server.storage import save_json, load_json

_store: dict[str, Game] = {}


def get_game(game_id: str) -> Game | None:
    return _store.get(game_id)


def set_game(game: Game) -> None:
    _store[game.id] = game


def get_or_load_game(game_id: str) -> Game | None:
    if game_id in _store:
        return _store[game_id]
    data = load_json(f"games/{game_id}/state.json")
    if data:
        game = Game.model_validate(data)
        _store[game_id] = game
        return game
    return None


def snapshot_game(game: Game) -> None:
    save_json(f"games/{game.id}/state.json", game.model_dump())
